### DISCUSSION TITLE
Usage with Svelte stores

### DISCUSSION THREAD

**Discussion by dpmccabe** [urql-graphql/urql#2113] [Q&A]
I have a moderately complex Svelte app using graphql+fetch and I was considering switching to urql. As in any complex Svelte app, you inevitably end up needing to share the same data between many components. Since this data doesn't necessarily belong to any particular hierarchy of components, the recommendation is to create [Svelte stores](https://svelte.dev/docs#svelte_store) (see also "Contexts vs. stores" on https://svelte.dev/tutorial/context-api), then import the stores in the components that need to access/modify their data.

This works well from an architecture standpoint: It separates your UI (components) from your data (stores) while retaining all of the reactivity you need. But this setup is going to be problematic for urql, which requires doing everything within Svelte contexts. Are there any example design patterns out there for using urql with dedicated Svelte stores? I really can't find anything except bare-bones "TODO list" kind of apps.

**Comment by jonathanstanley** [comment]
It appears to stem from https://github.com/FormidableLabs/urql/pull/1016 and may be mostly because the devs had less familiarity with svelte at the time? urql docs skew towards react, and the code appears to do the same.

As you point out, the dependence on `context` causes a number of problems:

- have to put all the logic into a component, or [go boom](https://github.com/FormidableLabs/urql/discussions/2110). (but as you point out, we may want the logic to be shared between multiple components)
- that issue means [quite a bit of duplicate code](https://github.com/FormidableLabs/urql/discussions/2201)
- it also [seems confusing](https://github.com/FormidableLabs/urql/discussions/2310) when svelte makes imports easy
- it looks more like `react` and not so much `svelte` ... which i think can be [hard to diagnose](https://github.com/FormidableLabs/urql/discussions/1988)
- [can't create tests](https://github.com/FormidableLabs/urql/discussions/2343)
- I wonder if these are slowing progress for [graphql-code-generator operation store plugin](https://github.com/dotansimha/graphql-code-generator/pull/6074)?

this issue seems to be at the core of many of the recent [urql discussions re: svelte](https://github.com/FormidableLabs/urql/discussions?discussions_q=svelte+) it almost seems like it could be easier to just skip the svelte bindings and use urql/core (or at least a documentation issue with this)... 

# Next steps
Perhaps the question should be: is Formidable open to change this or consider a PR to change this?

**Reply by frederikhors** [reply]
Dear guys,

`urql` is wonderful and the team that works there as well. They ousted Apollo (with good reason) and whoever uses it is satisfied.

It also works great with Svelte and I have been using it in two projects with success. Of course, the interaction should be improved and I am sure that as soon as they have time they will.

I am writing to report - to you but also to the `urql` team - two very interesting projects for Svelte & GraphQL:

1. https://github.com/jycouet/kitql, by the amazing @jycouet which also comes from `urql` for Svelte. Any contribution to the project is welcome;
1. https://github.com/HoudiniGraphql/houdini, by @AlecAivazis which is working on the new v2 cache.

The `urql` team is wonderful, I have been working with OSS for years and I have never seen a team so organized, fast and with such high code quality.

**Reply by jonathanstanley** [reply]
I also noticed @kitten comments in https://github.com/FormidableLabs/urql/issues/2195#issuecomment-1030390314 , notably:
> I'm not happy at all with our current API in @urql/svelte

> We don't work with Svelte outside of the urql repo. 

> We're at our limit in the amount of code we can effectively maintain

I think I'm going to have a look at offering this as a PR.  It may be much easier.

**Reply by jonathanstanley** [reply]
I have rough prototype for a query working locally which can be used in .svelte files and independently! 🚀  

With that working, I think getting the rest will be easier than I thought. I think it will solve the issues mentioned, but:

- This would be a breaking change from the current implementation ( I think ultimately easier to use and maintain)
- I would plan to comment my code 
- I would plan to provide corresponding update to the docs

Since a PR would take me a chunk of time, I would like some sense that the team would want it? @kitten  @JoviDeCroock ?

**Reply by JoviDeCroock** [reply]
Yes would be more than open to a PR! Really happy to hear someone wants to take this on ❤️

**Reply by jonathanstanley** [reply]
submitted PR: https://github.com/FormidableLabs/urql/pull/2370
